component/pprof/scrape/scrape.go

```diff
@@ -0,0 +1,310 @@
+					Delta:   true,
+					Path:    "/debug/pprof/profile",
+				},
+			},
+		},
+	}
+}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (arg *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*arg = NewDefaultArguments()
+
+	type args Arguments
+	if err := f((*args)(arg)); err != nil {
+		return err
+	}
+	if arg.ProfilingConfig == nil || arg.ProfilingConfig.PprofConfig == nil {
+		arg.ProfilingConfig = DefaultArguments.ProfilingConfig
+	} else {
+		for pt, pc := range DefaultArguments.ProfilingConfig.PprofConfig {
+			if arg.ProfilingConfig.PprofConfig[pt] == nil {
+				arg.ProfilingConfig.PprofConfig[pt] = pc
+				continue
+			}
+			if arg.ProfilingConfig.PprofConfig[pt].Enabled == nil {
+				arg.ProfilingConfig.PprofConfig[pt].Enabled = trueValue()
+			}
+			if arg.ProfilingConfig.PprofConfig[pt].Path == "" {
+				arg.ProfilingConfig.PprofConfig[pt].Path = pc.Path
+			}
+		}
+	}
+
+	if arg.ScrapeTimeout > arg.ScrapeInterval {
+		return fmt.Errorf("scrape timeout must be larger or equal to inverval")
+	}
+	if arg.ScrapeTimeout == 0 {
+		arg.ScrapeTimeout = arg.ScrapeInterval
+	}
+
+	if cfg, ok := arg.ProfilingConfig.PprofConfig[pprofProcessCPU]; ok {
+		if *cfg.Enabled && arg.ScrapeTimeout < time.Second*2 {
+			return fmt.Errorf("%v scrape_timeout must be at least 2 seconds", pprofProcessCPU)
+		}
+	}
+
+	return nil
+}
+
+// Component implements the pprof.scrape component.
+type Component struct {
+	opts component.Options
+
+	reloadTargets chan struct{}
+
+	mut        sync.RWMutex
+	args       Arguments
+	scraper    *Manager
+	appendable *pprof.Fanout
+}
+
+var _ component.Component = (*Component)(nil)
+
+// New creates a new pprof.scrape component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	flowAppendable := pprof.NewFanout(args.ForwardTo, o.ID, o.Registerer)
+	scraper := NewManager(flowAppendable, o.Logger)
+	c := &Component{
+		opts:          o,
+		reloadTargets: make(chan struct{}, 1),
+		scraper:       scraper,
+		appendable:    flowAppendable,
+	}
+
+	// Call to Update() to set the receivers and targets once at the start.
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	defer c.scraper.Stop()
+
+	targetSetsChan := make(chan map[string][]*targetgroup.Group)
+
+	go func() {
+		c.scraper.Run(targetSetsChan)
+		level.Info(c.opts.Logger).Log("msg", "scrape manager stopped")
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-c.reloadTargets:
+			c.mut.RLock()
+			var (
+				tgs     = c.args.Targets
+				jobName = c.opts.ID
+			)
+			if c.args.JobName != "" {
+				jobName = c.args.JobName
+			}
+			c.mut.RUnlock()
+			promTargets := c.componentTargetsToProm(jobName, tgs)
+
+			select {
+			case targetSetsChan <- promTargets:
+				level.Debug(c.opts.Logger).Log("msg", "passed new targets to scrape manager")
+			case <-ctx.Done():
```

Comment:
Do we need to return here if the context is done?

Reply:
It doesn't necessarily need to because it'll return on the next loop, but I think for clarity it would be good to return here anyway.